### ISSUE TITLE
Firefox Uncaught in promise

### ISSUE THREAD

**Issue by devnulling** [maia-sdr/maia-sdr#2]
Ran into this issue trying to use Firefox with a Linux host.

This was a fatal error on FF93 (really old I know :) ) where the spectrum would not display. Updating to FF 109 (most recent version) it is no longer fatal and the spectrum displays but throws a slightly different error.  I'm including the FF93 output as it might be a bit more helpful, the FF 109 doesn't output very much.

FF 109 non-fatal error:

```
Uncaught (in promise) unable to parse maia_json::Ad9361 JSON 
```

FF 93 error:

```
Uncaught (in promise) 
<dialog id="recording_dialog" class="ui">
​
accessKey: ""
​
accessKeyLabel: ""
​
assignedSlot: null
​
attributes: NamedNodeMap [ class="ui", id="recording_dialog" ]
​
baseURI: "http://192.168.2.1:8000/"
​
childElementCount: 1
​
childNodes: NodeList(3) [ #text, form#recording_form, #text ]
​
children: HTMLCollection { 0: form#recording_form, length: 1, … }
​
classList: DOMTokenList [ "ui" ]
​
className: "ui"
​
clientHeight: 0
​
clientLeft: 0
​
clientTop: 0
​
clientWidth: 0
​
contentEditable: "inherit"
​
dataset: DOMStringMap(0)
​
dir: ""
​
draggable: false
​
firstChild: #text "
      "
​
firstElementChild: <form id="recording_form" method="dialog">
​
hidden: false
​
id: "recording_dialog"
​
innerHTML: "\n      <form method=\"dialog\" id=\"recording_form\">\n        <label for=\"recording_metadata_file\">Filename</label>\n        <input type=\"text\" id=\"recording_metadata_filename\">\n        <label for=\"recording_metadata_description\">Description</label>\n        <input type=\"text\" id=\"recording_metadata_description\">\n        <label for=\"recording_metadata_author\">Author</label>\n        <input type=\"text\" id=\"recording_metadata_author\">\n        <label for=\"recorder_mode\">Mode</label>\n        <select id=\"recorder_mode\">\n          <option>8 bit IQ</option>\n          <option>12 bit IQ</option>\n        </select>\n        <a id=\"download_recording\" href=\"/recording\" download=\"\">Download recording</a>\n        <button id=\"close_recording_dialog\" value=\"close\" autofocus=\"\">Close</button>\n      </form>\n    "
​
innerText: "\n      \n        Filename\n        \n        Description\n        \n        Author\n        \n        Mode\n        \n          8 bit IQ\n          12 bit IQ\n        \n        Download recording\n        Close\n      \n    "
​
isConnected: true
​
isContentEditable: false
​
lang: ""
​
lastChild: #text "
    "
​
lastElementChild: <form id="recording_form" method="dialog">
​
localName: "dialog"
​
namespaceURI: "http://www.w3.org/1999/xhtml"
​
nextElementSibling: <canvas id="canvas" style="cursor: crosshair;">
​
nextSibling: #text "

    "
​
nodeName: "DIALOG"
​
nodeType: 1
​
nodeValue: null
​
nonce: ""
​
offsetHeight: 0
​
offsetLeft: 0
​
offsetParent: null
​
offsetTop: 0
​
offsetWidth: 0
​
onabort: null
​
onanimationcancel: null
​
onanimationend: null
​
onanimationiteration: null
​
onanimationstart: null
​
onauxclick: null
​
onbeforeinput: null
​
onblur: null
​
oncanplay: null
​
oncanplaythrough: null
​
onchange: null
​
onclick: null
​
onclose: null
​
oncontextmenu: null
​
oncopy: null
​
oncuechange: null
​
oncut: null
​
ondblclick: null
​
ondrag: null
​
ondragend: null
​
ondragenter: null
​
ondragexit: null
​
ondragleave: null
​
ondragover: null
​
ondragstart: null
​
ondrop: null
​
ondurationchange: null
​
onemptied: null
​
onended: null
​
onerror: null
​
onfocus: null
​
onformdata: null
​
onfullscreenchange: null
​
onfullscreenerror: null
​
ongotpointercapture: null
​
oninput: null
​
oninvalid: null
​
onkeydown: null
​
onkeypress: null
​
onkeyup: null
​
onload: null
​
onloadeddata: null
​
onloadedmetadata: null
​
onloadend: null
​
onloadstart: null
​
onlostpointercapture: null
​
onmousedown: null
​
onmouseenter: null
​
onmouseleave: null
​
onmousemove: null
​
onmouseout: null
​
onmouseover: null
​
onmouseup: null
​
onmozfullscreenchange: null
​
onmozfullscreenerror: null
​
onpaste: null
​
onpause: null
​
onplay: null
​
onplaying: null
​
onpointercancel: null
​
onpointerdown: null
​
onpointerenter: null
​
onpointerleave: null
​
onpointermove: null
​
onpointerout: null
​
onpointerover: null
​
onpointerup: null
​
onprogress: null
​
onratechange: null
​
onreset: null
​
onresize: null
​
onscroll: null
​
onsecuritypolicyviolation: null
​
onseeked: null
​
onseeking: null
​
onselect: null
​
onselectionchange: null
​
onselectstart: null
​
onslotchange: null
​
onstalled: null
​
onsubmit: null
​
onsuspend: null
​
ontimeupdate: null
​
ontoggle: null
​
ontransitioncancel: null
​
ontransitionend: null
​
ontransitionrun: null
​
ontransitionstart: null
​
onvolumechange: null
​
onwaiting: null
​
onwebkitanimationend: null
​
onwebkitanimationiteration: null
​
onwebkitanimationstart: null
​
onwebkittransitionend: null
​
onwheel: null
​
outerHTML: "<dialog class=\"ui\" id=\"recording_dialog\">\n      <form method=\"dialog\" id=\"recording_form\">\n        <label for=\"recording_metadata_file\">Filename</label>\n        <input type=\"text\" id=\"recording_metadata_filename\">\n        <label for=\"recording_metadata_description\">Description</label>\n        <input type=\"text\" id=\"recording_metadata_description\">\n        <label for=\"recording_metadata_author\">Author</label>\n        <input type=\"text\" id=\"recording_metadata_author\">\n        <label for=\"recorder_mode\">Mode</label>\n        <select id=\"recorder_mode\">\n          <option>8 bit IQ</option>\n          <option>12 bit IQ</option>\n        </select>\n        <a id=\"download_recording\" href=\"/recording\" download=\"\">Download recording</a>\n        <button id=\"close_recording_dialog\" value=\"close\" autofocus=\"\">Close</button>\n      </form>\n    </dialog>"
​
ownerDocument: HTMLDocument http://192.168.2.1:8000/
​
parentElement: <body>
​
parentNode: <body>
​
part: DOMTokenList []
​
prefix: null
​
previousElementSibling: null
​
previousSibling: #text "

    "
​
scrollHeight: 0
​
scrollLeft: 0
​
scrollLeftMax: 0
​
scrollTop: 0
​
scrollTopMax: 0
​
scrollWidth: 0
​
shadowRoot: null
​
slot: ""
​
spellcheck: false
​
style: CSS2Properties(0)
​
tabIndex: -1
​
tagName: "DIALOG"
​
textContent: "\n      \n        Filename\n        \n        Description\n        \n        Author\n        \n        Mode\n        \n          8 bit IQ\n          12 bit IQ\n        \n        Download recording\n        Close\n      \n    "
​
title: ""
​
<prototype>: HTMLUnknownElementPrototype { … }


```

**Comment by daniestevez** [comment]
I haven't tested with older browsers, but I'm not aware of a feature that I'm using that wasn't available in FF93 (WebGL2 support appeared in FF51, for instance). The FF93 error is quite verbose, but not too helpful for me. I can't see which promise triggered the problem. The error seems a dump of the internal properties of the `<dialog>` HTML element used to enter the recording metadata. I don't know why this is dumped.

I'm using FF109 and I've just noticed that I also get the `Uncaught (in promise) unable to parse maia_json::Ad9361 JSON` error. The reason why this happens can be seen if you run the web developer tools and refresh the page. There is a PATCH request to `api/ad9361` that fails with the error `"failed to set IIO attribute rx_gain: Operation not supported (os error 95)"`. Server-side errors are reported with an HTTP code 500 and an error message in JSON. The web app is nevertheless trying to parse the JSON as if it had the usual JSON format that an HTTP code 200 would produce (in this case a JSON containing the AD9361 state). So this is where the error happens. An obvious improvement would be:

* [ ] maia-wasm should look at the HTTP status code before trying to parse the JSON output from an API request.

I don't know what would be best to do if the HTTP status code indicates an error. Perhaps we can log an error to the console using the error message in the JSON instead of having the promise fail. That would be more informative for the user.

The underlying problem here is the following. When the web app is loaded, it reads the "preferences" that might have been saved to local storage earlier (this is basically the state of the UI, including RX frequency, sampling rate, etc, and is a way of having the settings persist across reboots of the Pluto because your browser remembers them). For each setting, there is a PATCH request to `api/ad9361` to update the value on the server. The problem is that the previous PATCH has set `rx_gain_mode: "SlowAttack"`, and the PATCH that fails is trying to set `rx_gain: 71`. The AD9361 sysfs interface refuses to set the `rx_gain` when the AGC mode is not "manual". So this is where the error actually originates.

Therefore, the error is harmless, because the preferences have been applied in the server in the best way possible (everything except for the `rx_gain`, which doesn't make sense with automatic gain, has been applied successfully).

* [ ] Perhaps the web app should be smarter and not try to set the `rx_gain` when the `rx_gain_mode` is not `"Manual"`. 
Currently the web app doesn't know anything about the relationship between these two parameters.

**Comment by daniestevez** [comment]
I've merged #6, which implements the two changes I proposed for this. I think this issue can be closed now (feel free to close it).

**Comment by daniestevez** [comment]
Closing, since this was fixed in #6.
